### ISSUE TITLE
I want to try to make a Mac OS version

### ISSUE THREAD

**Issue by ebellumat** [yuzu-emu/yuzu#6527]
Hi, i want to try a port of this emulator to run on Mac OS.

Theres any advice that can help me with this achievement? 

I Know that try to make this to run on Metal API because Mac OS can't run Vulkan can be challenge.

**Comment by athei** [comment]
The Vulkan SDK does support macOS via moltenVK. However, it is missing some features that are used by yuzu.

I would tackle this task in two separate PRs:

1. Make yuzu **compile** on the macOS native Xcode (clang) compiler. Please note that is lags behind the upstream clang but a new one will be released end of Summer when new the OS ships. Main offenders here are bleeding Edge C++20 features (concepts and ranges). Maybe you can use conditional compilation or some compat libs to support those features on clang. Please note that this won't add Vulkan (via MoltenVK) but merely fixes the compilation. This would also include changes to the cmake in order to create a proper application bundle (I think the current one does not bundle all the required libs). A CI job would be great to to make sure compilation won't break with the next PR.

2. Add VulkanSDK on macOS.

I would try to tackle the first one and get it merged before even thinking about Metal/Vulkan.

**Comment by ebellumat** [comment]
Thank you @athei, this is a very good roadmap, makes sense.

**Comment by athei** [comment]
Take it with a grain of salt. I am not a contributor here and cannot tell whether such a clang compat PR would be accepted.

**Comment by MerryMage** [comment]
See also issues #6038, #6042, #6049.

**Comment by a727990982** [comment]
can m1 mac use this yuzu?i see some person has run it on m1.

**Comment by athei** [comment]
It can't be run on Intel Macs right now because of the mentioned stuff (clang, MoltenVK) so it won't run on M1 Macs either. Once it runs on Intel Macs it **should** run under Rosetta. AFAIK Rosetta can also handle JITTED Intel code. To run it natively a M1 backend needs to be added to dynamic. Since both source and target use ARM that might a a really fast JIT.

**Comment by Morph1984** [comment]
Do note that we've also transitioned to `std::filesystem` and started using `char8_t` in #6270. Hopefully these aren't additional barriers to getting macOS support working

**Comment by Morph1984** [comment]
Since `char8_t` is the only portable type which is defined for storing UTF-8 code units across platforms, I have used it instead of `char` and `std::filesystem::u8path`

**Comment by athei** [comment]
Those are supported in clang. But this is why we need a CI job for the current Xcode used clang version as soon as the initial port is finished. Otherwise unsupported features slip through the cracks pretty easily.

**Comment by iMonZ** [comment]
> The Vulkan SDK does support macOS via moltenVK. However, it is missing some features that are used by yuzu.
> 
> I would tackle this task in two separate PRs:
> 
> 1. Make yuzu **compile** on the macOS native Xcode (clang) compiler. Please note that is lags behind the upstream clang but a new one will be released end of Summer when new the OS ships. Main offenders here are bleeding Edge C++20 features (concepts and ranges). Maybe you can use conditional compilation or some compat libs to support those features on clang. Please note that this won't add Vulkan (via MoltenVK) but merely fixes the compilation. This would also include changes to the cmake in order to create a proper application bundle (I think the current one does not bundle all the required libs). A CI job would be great to to make sure compilation won't break with the next PR.
> 2. Add VulkanSDK on macOS.
> 
> I would try to tackle the first one and get it merged before even thinking about Metal/Vulkan.

Actually, it could be possible to release a macOS version after the two issues are fixed.
And if some new features gonna break with the new yuzu version, you can still keep the old macOS version until it's fixed.
Sounds like a good approach?
Yuzu would be so great with macOS support, even when it's not being updated frequently.

**Comment by athei** [comment]
> Sounds like a good approach?

No. The important part is that `master` compiles on all supported operating systems at all times. Otherwise no serious development activity is possible by macOS developers.

**Comment by iMonZ** [comment]
> > Sounds like a good approach?
> 
> No. The important part is that `master` compiles on all supported operating systems at all times. Otherwise no serious development activity is possible by macOS developers.

That's true, but If the project compiles like every two months, this would be better than no version at all.
It depends on how much timer the developers have.

**Comment by surajmandalcell** [comment]
Is this issue still active?

**Comment by ericcurtin** [comment]
For M1, a port to aarch64 Linux would make sense to start, Linux VMs work great on M1 and native support for Linux should be stable by the time an aarch64 port was complete https://ericcurtin.wordpress.com/2022/01/14/more-than-a-pi-alternative-aarch64-environments/

**Comment by ebellumat** [comment]
Makes sense to me, @ericcurtin

**Comment by NickStrupat** [comment]
Should be able to use https://github.com/KhronosGroup/MoltenVK to enable Vulkan support in a macOS build.

Xcode now comes with clang 13.0.0.

**Comment by iMonZ** [comment]
> Makes sense to me, @ericcurtin, i think, this big challenge to port any 3D stuff at Apple ecosystem is the lack of Vulkan support. Metal is great, but is awafull to write something just for one platform.

You have MoltenVK and Zink.
And if apple updates their Xcode to modern library’s (like they did 4 months ago it could even Work)

**Comment by ronpetit** [comment]
> ~Xcode now comes with clang 13.0.0.~ Xcode clang 13 is still not close enough to mainline clang 13.
> 
> It seems the easiest way to get mainline clang 13 on macOS is to install via `homebrew`

As @athei stated we should use the macOS's default clang, which is the mainline version we exactly need to compile yuzu?, I have the lastest xcode version and when I do a `clang --version` on terminal I got this:

```bash
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: x86_64-apple-darwin21.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by ericcurtin** [comment]
Target x86_64 mac? Just wondering why not favour a port to aarch64 (starting with Linux if it makes it easier)? Rather than a platform that's being retired...

**Comment by ronpetit** [comment]
> Target x86_64 mac? Just wondering why not favour a port to aarch64 (starting with Linux if it makes it easier)? Rather than a platform that's being retired...

@ericcurtin I'm just wondering if the latest clang around macOS is the one that @NickStrupat refers

And about favoring x86_64 or aarch64, I think that depends on what the Mac developers have right now, the M1 Pro and M1 Max are new and most developers with high end devices are still using x86_64. I want to switch, but there are a couple of SDKs that are not on aarch64 yet (Android Studio emulators for ARM took a while to arrive)

I think it will take a while, maybe 1 or 2 years to get around 50% adoption of ARM devices for developers, but that's an opinion

**Comment by ramzzzay** [comment]
I'm here with my m1 pro 2021 MacBook pro. Can I maybe help to you @ronpetit with testing builds or something else on arm64 arch?

```
clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2)
Target: arm64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by ericcurtin** [comment]
I still think the baby steps approach might be easier just porting to aarch64 first... There's an asahilinux installer, a debian installer and many VM options for M1 Mac now.

**Comment by jefer94** [comment]
I listen about a developer port Yuzu to M1, that's true? I thinking about buy a Mac Studio and in the list of software I check is Yuzu in the list, my question is how go the progress about the port of if is easy compile at least a build very buggy, I in the pass I was compile CyanogenMod and Linux

**Comment by jefer94** [comment]
More info about the progress here https://github.com/yuzu-emu/yuzu/pull/4165

**Comment by JuniorJPDJ** [comment]
Asahi still doesn't support GPU, so it's not the best testing idea - VM with linux on Parallels has GPU accel on aarch64.

**Comment by iMonZ** [comment]
> For M1, a port to aarch64 Linux would make sense to start, Linux VMs work great on M1 and native support for Linux should be stable by the time an aarch64 port was complete https://ericcurtin.wordpress.com/2022/01/14/more-than-a-pi-alternative-aarch64-environments/



> Asahi still doesn't support GPU, so it's not the best testing idea - VM with linux on Parallels has GPU accel on aarch64.

Asahi doesn’t support gpu and thus no Wayland.
I think we can forget Asahi until 2024 probably even when the progress is insane!

**Comment by KohlerVG** [comment]
Have you seen the new Metal 3 API Apple released? It has far better Raytracing features than before: https://developer.apple.com/metal/

**Comment by LuigiFan25** [comment]
It also has shaders like geometry shaders that should help a lot

**Comment by ebellumat** [comment]
The Geometry Shaders would be great for the Wine/Crossover. I'm unable to run Unreal Engine 4 Windows games on Mac because the lack of Geometry Shaders support on Metal 2.

**Comment by czkboy** [comment]
what about metal3

**Comment by ericcurtin** [comment]
> 

s/2024/2022/g it sure does support GPU!

**Comment by ericcurtin** [comment]
> > For M1, a port to aarch64 Linux would make sense to start, Linux VMs work great on M1 and native support for Linux should be stable by the time an aarch64 port was complete https://ericcurtin.wordpress.com/2022/01/14/more-than-a-pi-alternative-aarch64-environments/
> 
> > Asahi still doesn't support GPU, so it's not the best testing idea - VM with linux on Parallels has GPU accel on aarch64.
> 
> Asahi doesn’t support gpu and thus no Wayland. I think we can forget Asahi until 2024 probably even when the progress is insane!

s/2024/2022/g it sure does support GPU!

**Comment by bill-7** [comment]
https://blog.ryujinx.org/the-impossible-port-macos/

time to turn to the dark side, friends!

**Comment by MahdeenSky** [comment]
> 

massive achievement, i applaud their work.

**Comment by ericcurtin** [comment]
If anybody want access to an early version of the Fedora Asahi distro to try and port yuzu feel free to get in touch. There's a little bit of polishing going on at the moment (mainly logos and stuff) and I expect official release to be within the next 2 months.
